### PR TITLE
Workaround for helm3 compatibility

### LINKED ISSUE #943: Helm deployment fails with helm3 when "deploying clusterStacks with helm" enabled
### What happened?

The "cluster stack installation with Crossplane helm chart" feature introduced with #932 is not working with helm 3.

### How can we reproduce it?

Run the following to deploy Crossplane helm chart with helm3 enabling at least one clusterStack:
```bash
$ helm3 install crossplane-test --namespace crossplane-system -f values.yaml . --set clusterStacks.aws.deploy=true
Error: apiVersion "stacks.crossplane.io/v1alpha1" in crossplane/templates/clusterstackinstalls.yaml is not available
```

### What environment did it happen in?
Crossplane version: master including #932 
Helm version: v3.0.0-beta.4

### PR DESCRIPTION
**Pull request by turkenh**
### Description of your changes
Helm 3 introduced some breaking changes one of which is [deprecating `crd-install` hook](https://helm.sh/docs/topics/charts_hooks/#the-available-hooks) and skipping CRDs with this annotation which causes some of the [existing helm charts to fail](https://github.com/helm/charts/issues/9597). We need these hooks for `stackinstall` and `clusterstackinstall` CRDs to be able to [install infrastructure stacks with helm](https://github.com/crossplaneio/crossplane/pull/932). 

Ideally we should drop helm2 support and specify helm3 as minimum version as suggested by @soorena776 [here](https://github.com/crossplaneio/crossplane/issues/1062) as potential fix. However, there are still users (like [gitlab](https://gitlab.com/groups/gitlab-org/-/epics/1866#note_227663961)) who depend on helm2 and we need to support it at least for some more time.

One workaround could be using some sort of built-in variable which lets us to understand it is helm 3 and we do not put annotation if so, however [there seems to be no such variable yet](https://github.com/helm/helm/issues/6689).

As a workaround, this PR copies the two crds, without `crd-install` hook annotation, under `<chart>/crds` directory which is introduced with helm3 but ignored in helm2. 

So, with this change, we will have `clusterstackinstalls` and `stackinstalls` CRD yamls:

- with hook under `<crossplane-chart>/templates/crds`: for `helm2` skipped in `helm3`
- without hook under `<crossplane-chart>/crds`: for `helm3` ignored in `helm2`

Fixes #943, #1062, #1064 
### Checklist

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the RBAC permissions in [`clusterrole.yaml`] to include any new types.

[documentation]: https://github.com/crossplaneio/crossplane/tree/master/docs
[examples]: https://github.com/crossplaneio/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplaneio/crossplane/tree/master/PendingReleaseNotes.md
[`clusterrole.yaml`]: https://github.com/crossplaneio/crossplane/blob/master/cluster/charts/crossplane/templates/clusterrole.yaml

### CODE REVIEW

**Comment by soorena776** [comment]
@turkenh Thanks for the quick fix! As for backward compatibility, my understanding is that Gitlab is using an already tagged version of Crossplane (e.g. v0.4), so not supporting helm2 going forward shouldn't break their existing flow. Isn't that the case @prasek and @jbw976?

File: Makefile [hunk 1]
```diff
@@ -83,7 +83,18 @@ manifests:
 
 generate: $(KUSTOMIZE) go.generate manifests.annotate
 
-
+CROSSPLANE_CHART_DIR=cluster/charts/crossplane
+CRD_DIR = $(CROSSPLANE_CHART_DIR)/templates/crds
+HELM3_CRD_DIR = $(CROSSPLANE_CHART_DIR)/crds
+manifests.annotate:
```
Comment (displague) [inline Makefile]: nitpic about spacing and flow..

The comment block that was above manifests.annotate should stay above the `manifests.annotate:` variables.  Your new comments should be incorporated into that comment block.

https://github.com/crossplaneio/crossplane/pull/1077/files#diff-b67911656ef5d18c4ae36cb6741b7965R86 should match the spacing of the other two variables.

**Comment by jbw976** [comment]
@soorena776 I thought GitLab managed apps was using whatever the latest is from the alpha charts repo, but you might be right that they are actually always using v0.4.1 from the alpha repo.  I'm confirming with @mahendrabagul in this note on the MR: https://gitlab.com/gitlab-org/gitlab/merge_requests/18797/diffs#note_245343086

File: Makefile (hunk 1 above)
Comment (turkenh) [reply Makefile]: Thanks, both should be done now.

**Comment by hasheddan** [comment]
@turkenh any update on this PR? Does is need further review?

**Comment by turkenh** [comment]
@hasheddan PR is ready from my POV.

But from the discussions above it looks like we can also drop helm 2 support and adapt the chart only for helm 3, so we wouldn't need this workaround. @jbw976, I see that Gitlab using a hardcoded version of  Crossplane Helm chart, does this mean we can drop Helm 2 support? Wouldn't they need to update Crossplane version to get new feature and bug fixes? 

In any case, my suggestion would be to merge this workaround ASAP and follow dropping Helm 2 support with [another issue](https://github.com/crossplaneio/crossplane/issues/1100).

**Comment by jbw976** [comment]
>  @jbw976, I see that Gitlab using a hardcoded version of Crossplane Helm chart, does this mean we can drop Helm 2 support? Wouldn't they need to update Crossplane version to get new feature and bug fixes?
Right, in GitLab I imagine they will update to v0.5 and v0.6 over time, which could very well be before GitLab moves to Helm 3.  I think we need to maintain compatibility with Helm 2 for an undetermined amount of time going forward still.

File: Makefile [hunk 2]
```diff
@@ -96,8 +95,18 @@ generate: $(KUSTOMIZE) go.generate manifests.annotate
 # policy is necessary to be able to redeploy helm chart after it is deleted
 # since CRDs with "crd-install" hooks will not be deleted with "helm delete" and
 # cause next "helm install" to fail.
-CRD_DIR = cluster/charts/crossplane/templates/crds
+# "helm.sh/hook: crd-install" deprecated in helm3 and crds with this annotation
```
Comment (jbw976) [inline Makefile]: Could this comment be cleaned up and explained just a bit more?  e.g.,

```make
# "helm.sh/hook: crd-install" was deprecated in helm3 and CRDs with this annotation are
# skipped. This results in the StackInstall and ClusterStackInstall CRDs not being
# installed in helm3 when they have that annotation.
# As a workaround, we first copy those CRDs under <chart>/crds directory which
# was introduced with helm3 and ignored in helm2, then afterwards apply the annotation to
# those CRDs under <chart>/templates/crds for helm2.
```

Comment (turkenh) [reply Makefile]: Done! Thanks for the suggested text.

**Comment by turkenh** [comment]
> I assume you tested this for both helm2 and helm3?

Yes, tested with both versions (v2.15.1 & v3.0.0) with/without cluster stacks.

> On helm3, is the <chart>/templates/crds directory ignored? How does helm3 know not to just render everything under <chart>/templates?

No, helm3 already renders everything under `<chart>/templates/crds` but skips the two with `crd-install` hook. Those two comes from `<chart>/crds` as helm3 recognizes this directory.

**Comment by infinitecompute** [comment]
When this is merged, what will the end state be regarding installation from alpha when using either Helm 2 or 3?  Will the docs need a new section describing how to install with Helm 3 changes (e.g. generating the namespace prior to running helm, and how to get pull down the workaround for <chart>/templates/crds)?  When I do `brew install helm`, I am now on helm 3 by default, so I have to decide whether to pin myself to the older version.

**Comment by turkenh** [comment]
> When this is merged, what will the end state be regarding installation from alpha when using either Helm 2 or 3? Will the docs need a new section describing how to install with Helm 3 changes (e.g. generating the namespace prior to running helm?

There should be no difference regarding installing with helm 2 or helm 3, except:
- With helm3, namespace to be created prior to install as you pointed out, but I see it is already documented [here](https://github.com/crossplaneio/crossplane/blob/master/docs/install-crossplane.md#installation).
- With helm3, Following two lines of log which would be totally harmless:
```
manifest_sorter.go:175: info: skipping unknown hook: "crd-install"
manifest_sorter.go:175: info: skipping unknown hook: "crd-install"
```

**Pull request merged**